fix: Migrate OnClientConnectedCallback invocation into StartHost [MTT-4972]

NoelStephensUnity opened the pull request:
This assures that any NetworkBehaviour components attached to in-scene placed NetworkObjects will have their associated netcode/NetworkManager properties properly initialized when OnClientConnected is invoked.

[MTT-4972](https://jira.unity3d.com/browse/MTT-4972)

## Changelog
- Fixed: The issue where `NetworkManager.OnClientConnectedCallback` was being invoked before in-scene placed `NetworkObject`s had been spawned when starting `NetworkManager` as a host.

## Testing and Documentation
- Includes integration test (wip)

NoelStephensUnity committed:
fix

migrate InvokeOnClientConnectedCallback for host only.

NoelStephensUnity committed:
style

adding comment

JesseOlmer commented:
Is there a bug # this is associated with?

HandleConnectionApproval just above in the same method invokes onclientconnected... either directly if scene management is disabled, or at the end of the Synchronize scene event if scene management is enabled.

Isn't this change just going to fire it multiple times?

NoelStephensUnity commented:
> Is there a bug # this is associated with?
> 
> HandleConnectionApproval just above in the same method invokes onclientconnected... either directly if scene management is disabled, or at the end of the Synchronize scene event if scene management is enabled.
> 
> Isn't this change just going to fire it multiple times?

Yeah, I forgot to mark it as draft.  It was just a quick first pass example of what it would sort of look like.
But you are right about when scene management is disabled, that line needs to be deleted too.
It will not get invoked within NetworkSceneManager during a Synchronize scene event as that is only invoked on clients and not on the server-side (server is already synchronized).

NoelStephensUnity committed:
update

MTT-4972
Added the fixed entry for this PR.

NoelStephensUnity committed:
test

MTT-4972
adding validation test to existing NetworkManagerTests

NoelStephensUnity committed:
test

MTT-4972
Updating the integration test to validate that this update does not impact anything when scene management is disabled.
Unloading the loaded scene.

JesseOlmer commented:
Host will hit this, will it not?
https://github.com/Unity-Technologies/com.unity.netcode.gameobjects/blob/develop/com.unity.netcode.gameobjects/Runtime/Core/NetworkManager.cs#L2222

NoelStephensUnity commented:
It will not hit that line, in the PR branch that is, as it was
removed/migrated to StartHost:
https://github.com/Unity-Technologies/com.unity.netcode.gameobjects/blob/3439f11c6d1e70eea3e07864c8ce8efa9a3fc9f8/com.unity.netcode.gameobjects/Runtime/Core/NetworkManager.cs#L2222

On Wed, Oct 26, 2022 at 10:10 AM Jesse Olmer ***@***.***>
wrote:

> Host will hit this, will it not?
>
> https://github.com/Unity-Technologies/com.unity.netcode.gameobjects/blob/develop/com.unity.netcode.gameobjects/Runtime/Core/NetworkManager.cs#L2222
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Unity-Technologies/com.unity.netcode.gameobjects/pull/2277#issuecomment-1292199300>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AROMJ5P27CBNGPOYXSGVS4TWFFCWPANCNFSM6AAAAAAROKGR2E>
> .
> You are receiving this because you authored the thread.Message ID:
> <Unity-Technologies/com.unity.netcode.gameobjects/pull/2277/c1292199300@
> github.com>
>
